lib/OpenQA/WebAPI/Controller/Step.pm

```diff
@@ -91,6 +94,45 @@ sub view ($self) {
     $self->viewimg;
 }
 
+sub _checkout_needledir_or_default {
+    my $job = shift;
+    my $tmp_needles_dir = $job->needle_dir;
+    my $real_needle_dir = realpath($tmp_needles_dir) // $tmp_needles_dir;
+    if (my $casedir = $job->settings->single({key => 'CASEDIR'})) {
+        my $casedir_refspec = Mojo::URL->new($casedir->value)->fragment;
+        eval {
+            my $vars_json = Mojo::File->new($job->result_dir(), 'vars.json')->slurp;
+            my $vars = decode_json($vars_json);
+            $casedir_refspec = $vars->{TEST_GIT_HASH};
+        };
```

Comment:
I suppose the case when we cannot determine `$casedir_refspec` should be handled as soon as possible:

```suggestion
        };
        return $real_needle_dir if !$casedir_refspec;
```

Otherwise there's just no need to invoke `git` at all.